xclim/indices/generic.py

```diff
@@ -57,31 +59,120 @@
+      (along dimension 'time'), but with masked (NaN) values outside the period of
+      interest.
+
+    Examples
+    --------
+    Keep only the values of fall and spring.
+
+    >>> ds = open_dataset("ERA5/daily_surface_cancities_1990-1993.nc")
+    >>> ds.time.size
+    1461
+    >>> out = select_time(ds, drop=True, season=['MAM', 'SON'])
+    >>> out.time.size
+    732
+
+    Or all values between two dates (included).
+
+    >>> out = select_time(ds, drop=True, date_bounds=('02-29', '03-02'))
+    >>> out.time.values
+    array(['1990-03-01T00:00:00.000000000', '1990-03-02T00:00:00.000000000',
+           '1991-03-01T00:00:00.000000000', '1991-03-02T00:00:00.000000000',
+           '1992-02-29T00:00:00.000000000', '1992-03-01T00:00:00.000000000',
+           '1992-03-02T00:00:00.000000000', '1993-03-01T00:00:00.000000000',
+           '1993-03-02T00:00:00.000000000'], dtype='datetime64[ns]')
     """
-    if not indexer:
-        selected = da
-    else:
-        key, val = indexer.popitem()
-        time_att = getattr(da.time.dt, key)
-        selected = da.sel(time=time_att.isin(val)).dropna(dim="time")
+    N = sum(arg is not None for arg in [season, month, doy_bounds, date_bounds])
+    if N > 1:
+        raise ValueError(f"Only one method of indexing may be given, got {N}.")
+
+    if N == 0:
+        return da
+
+    if season is not None:
+        if isinstance(season, str):
+            season = [season]
+        mask = da.time.dt.season.isin(season)
+
+    elif month is not None:
+        if isinstance(month, int):
+            month = [month]
+        mask = da.time.dt.month.isin(month)
+
+    elif doy_bounds is not None:
+        start, end = doy_bounds
+        if start <= end:
+            doys = np.arange(start, end + 1)
+        else:
+            doys = np.concatenate((np.arange(start, 367), np.arange(0, end + 1)))
+        mask = da.time.dt.dayofyear.isin(doys)
+
+    elif date_bounds is not None:
+        # This one is a bit trickier.
+        start, end = date_bounds
+        time = da.time
+        calendar = get_calendar(time)
+        if calendar not in uniform_calendars:
+            # For non-uniform calendars, we can't simply convert dates to doys
+            # conversion to all_leap is safe for all non-uniform calendar as it doesn't remove any date.
+            time = convert_calendar(time, "all_leap")
+            # values of time are the _old_ calendar
+            # and the new calendar is in the coordinate
+            calendar = "all_leap"
+
+        # Get doy of date, this is now safe because the calendar is uniform.
+        start = to_cftime_datetime("2000-" + start, calendar).dayofyr
+        end = to_cftime_datetime("2000-" + end, calendar).dayofyr
+
+        if start <= end:
+            doys = np.arange(start, end + 1)
+        else:
+            doys = np.concatenate((np.arange(start, 367), np.arange(0, end + 1)))
+        mask = time.time.dt.dayofyear.isin(doys)
```

Comment:
Could you just set `doy_bounds` and move the `elif doy_bounds is not None:` code segment below to avoid repeating ?